### PR TITLE
[Ready for Review] Cane Upgrades for RP and Combat

### PR DESCRIPTION
**Pull request by BurgerLUA**
I found the cane to be annoying to use when it comes to doing actual work on the station. You can't use the cane to push buttons or generally interact with objects so I added something that would help.

But then I got carried away and added several other useful utility functions.

- Using help on someone with the cane will poke them, dealing 0.25 pain damage that usually regenerates within 4 seconds. Spamming poke with the cane can cause brute damage (takes like 5 in a row for it to happen), and can actually knock people down (RNG, rare after 10 pokes). Useful for characters playing old disgruntled men who like hitting youngsters with the cane without killing them in 5 hits.
- Using disarm on someone with the cane will perform an attack that takes 2.5 seconds to charge. If the arms are target, the weapon in that hand will be disarmed, else it will just deal 3 pain damage.
- Using grab on someone with the cane will perform a pull that takes 2.5 seconds to charge.  Equivalent to a normal Right Click > Pull however this one deals 1 pain damage on a successful pull.
- Using harm on someone with the cane will perform a regular attack that deals 10 brute damage to the targeted body part. 

PR is pulled as [WIP] as developers may not like the balancing around the cane. This also applies to any weapons that use the cane as a base, including the concealed cane (when concealed). If this PR is ever applied, I suggest setting the loadout points of the cane from 1 to 2, possibly 3.

### CODE REVIEW

**Comment by MarinaGryphon** [comment]
Setting it to three points to make up for it isn't really preferable, because canes are used by some characters for roleplay reasons other than combat. It would punish them unfairly for a mechanic they wouldn't be able to ICly make use of. It's definitely possible to nerf the new mechanics, but I'm not sure how desirable they are. I'll take a look at exactly what's added and provide more feedback later.

**Comment by BurgerLUA** [comment]
@MarinaGryphon 
I'll probably add some RNG to the cane's disarm, as that's the only mechanic I'm concerned about.

File: code/defines/obj/weapon.dm [hunk 1]
```diff
@@ -81,11 +81,78 @@
 	icon_state = "cane"
 	item_state = "stick"
 	flags = CONDUCT
-	force = 10
+	force = 0
 	throwforce = 7.0
 	w_class = 4
 	matter = list(DEFAULT_WALL_MATERIAL = 50)
-	attack_verb = list("bludgeoned", "whacked", "disciplined", "thrashed")
+	//attack_verb =
```
Comment (skull132) [inline code/defines/obj/weapon.dm]: Delete, do not comment.

File: code/defines/obj/weapon.dm [hunk 2]
```diff
@@ -81,11 +81,78 @@
...
+
+
+	var/blocked = H.run_armor_check(target_zone, "melee")
+	H.apply_damage(damageamount, damagetoapply, target_zone, blocked)
+	H.standard_weapon_hit_effects(src,user,damageamount,blocked,target_zone)
+	if (damagetoapply==BRUTE)
+		playsound(src.loc, "swing_hit", 50, 1, -1)
+	user.visible_message("<span class='[class]'>[user] [verbtouse] [target]'s [organname] with the [name][punct]</span>", "<span class='[class]'>You [verbtouse] [target]'s [organname] with the [name][punct]</span>")
+	user.do_attack_animation(H)
+	user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
```
Comment (skull132) [inline code/defines/obj/weapon.dm]: Missing `return 1`, you're breaking the attack chain effectively.

File: code/defines/obj/weapon.dm [hunk 3]
```diff
@@ -81,11 +81,78 @@
...
-	attack_verb = list("bludgeoned", "whacked", "disciplined", "thrashed")
+	//attack_verb =
+
+/obj/item/weapon/cane/afterattack(atom/A as mob|obj, mob/user as mob, proximity)
+	if (user.a_intent == I_HELP && !istype(A, /mob) && proximity)
+		A.attack_hand(user)
+
+/obj/item/weapon/cane/attack(mob/target as mob, mob/living/user as mob, var/target_zone)
+
+	var/mob/living/carbon/human/M = user
```
Comment (skull132) [inline code/defines/obj/weapon.dm]: This is not always true! Target specifically can be non-human, so the code should account for that. My suggestion is to run the old style attack code if the mob is non-human (a simpler mob):
```DM
/obj/item/weapon/cane/attack(mob/living/carbon/human/target, mob/living/carbon/human/user, target_zone)
	if (!istype(target) || !istype(user))
		return ..()

	// your code follows here.
```

File: code/defines/obj/weapon.dm [hunk 4]
```diff
@@ -81,11 +81,78 @@
 	icon_state = "cane"
 	item_state = "stick"
 	flags = CONDUCT
-	force = 10
+	force = 0
```
Comment (skull132) [inline code/defines/obj/weapon.dm]: Is there a particular reason for this? You're already overriding attack(), and there are other pieces of code which might expect a saner value.

Comment (LordFowl) [reply code/defines/obj/weapon.dm]: For example, this will make canes do no throw damage, I am sure.

**Comment by BurgerLUA** [comment]
Force is now back at 10. New attack damage is based on weapon's force so other, powerful, weapons can use this base.

Removed uncommented code.

I did the attack check wrong, fixing it now....

Added a return 1 to the end.

**Comment by BurgerLUA** [comment]
Forgot to mention that I also added some RNG to the code. Can be removed if necessary.

Chance to disarm while targeting the appropriate hand/arm is 50%. Targeting parts other than the hand will do a 25% chance per hand.

**Comment by BurgerLUA** [comment]
I posted this PR in the discord and the main concern is the disarm being useless because it takes so long. As it stands now, the code is fine (unless there are any objections) however the balance is the main concern since it's adding new and useful combat features to a weapon normally not used for combat.

**Comment by BurgerLUA** [comment]
Removed the WIP tag. Nothing else i can think of improving.

Delay for cane is 0.5 seconds, can be adjusted if too high or too low.

**Comment by BurgerLUA** [comment]
leaving as WIP again until travis cooperates

**Comment by BurgerLUA** [comment]
Removed WIP tag, can't think of anything off the top of my head that I could do to improve this.

File: code/defines/obj/weapon.dm [hunk 5]
```diff
@@ -87,6 +87,169 @@
 	matter = list(DEFAULT_WALL_MATERIAL = 50)
 	attack_verb = list("bludgeoned", "whacked", "disciplined", "thrashed")
 
+/obj/item/weapon/cane/attack(mob/living/target, mob/living/carbon/human/user, target_zone = "chest")
+
+	if(!(istype(target) && istype(user)))
+		return ..()
+
+	var/targetIsHuman = istype(target,/mob/living/carbon/human) // Thanks Lohikar for this check that I butchered
```
Comment (printer16) [inline code/defines/obj/weapon.dm]: Use `ishuman(target)` instead.

File: code/defines/obj/weapon.dm [hunk 6]
```diff
@@ -87,6 +87,169 @@
...
+	switch(user.a_intent)
+		if(I_HELP)
+			class = "notice"
+			punct = "."
+			soundname = 0
+			if (target_zone == "head" || target_zone == "eyes" || target_zone == "mouth")
+				verbtouse = pick("tapped")
+			else
+				verbtouse = pick("tapped","poked","prodded","touched")
+			damageamount = damageamount * 0.025
```
Comment (printer16) [inline code/defines/obj/weapon.dm]: `damageamount *= 0.025`

File: code/defines/obj/weapon.dm [hunk 7]
```diff
@@ -87,6 +87,169 @@
...
+								shoulddisarm = 2
+						else
+							if (prob(chancemod*0.5) && target.l_hand && target.l_hand != src)
+								shoulddisarm = 1
+							if (prob(chancemod*0.5) && target.r_hand && target.r_hand != src)
+								shoulddisarm += 2
+				else
+					user.visible_message("<span class='[class]'>[user] flips [user.get_pronoun(1)] [name] back to it's original position.</span>", "<span class='[class]'>You flip the [name] back to it's original position.</span>")
+					return 0
+			damageamount = damageamount * 0.25
```
Comment (printer16) [inline code/defines/obj/weapon.dm]: `damageamount *= 0.025`

File: code/defines/obj/weapon.dm [hunk 8]
```diff
@@ -87,6 +87,169 @@
...
+					if(!wasblocked && damageamount)
+						user.start_pulling(target)
+					else
+						verbtouse = pick("awkwardly tries to hook","fails to grab")
+				else
+					user.visible_message("<span class='[class]'>[user] flips [user.get_pronoun(1)] [name] back to it's original position.</span>", "<span class='[class]'>You flip the [name] back to it's original position.</span>")
+					return 0
+			else
+				soundname = "punch"
+			damageamount = damageamount * 0.1
```
Comment (printer16) [inline code/defines/obj/weapon.dm]: `damageamount *= 0.025`

File: html/changelogs/BurgerBB-CanePowercreep.yml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+author: BurgerBB
+
+delete-after: True
+
+changes: 
+  - rscadd: "Added new cane abilities based on intent. Help intent with a cane pokes people for 0.25 pain damage. Repeated poking will cause bruises and sometimes bleeding. Disarm intent with a cane smacks people for 2.5 pain damage, with a chance to disarm. Targeting the arm/hand will have a 50% chance to disarm that hand, targeting elsewhere has as a 25% chance to disarm the left hand, 25% chance to disarm the right hand. Grab intent with a cane grabs people and deals 1 pain damage, similiar to right clicking a target and pulling them. Harm intent with the cane deals 10 brute damage to the specific bodypart, with armor calculations."
```
Comment (Lohikar) [inline html/changelogs/BurgerBB-CanePowercreep.yml]: This is WAY too verbose.

**Comment by Lohikar** [comment]
WELL OKAY THEN

**Pull request merged**